universal-application-tool-0.0.1/app/controllers/admin/QuestionController.java

```diff
@@ -0,0 +1,113 @@
+import org.slf4j.LoggerFactory;
+import play.data.Form;
+import play.data.FormFactory;
+import play.mvc.Controller;
+import play.mvc.Http.Request;
+import play.mvc.Result;
+import services.question.InvalidPathException;
+import services.question.QuestionDefinition;
+import services.question.QuestionService;
+import services.question.UnsupportedQuestionTypeException;
+import views.admin.questions.QuestionEditView;
+import views.admin.questions.QuestionsListView;
+
+public class QuestionController extends Controller {
+  final Logger LOG = LoggerFactory.getLogger(this.getClass());
+
+  private final QuestionService service;
+  private final QuestionsListView listView;
+  private final QuestionEditView editView;
+  private final FormFactory formFactory;
+
+  @Inject
+  public QuestionController(
+      QuestionService service,
+      QuestionsListView listView,
+      QuestionEditView editView,
+      FormFactory formFactory) {
+    this.service = checkNotNull(service);
+    this.listView = checkNotNull(listView);
+    this.editView = checkNotNull(editView);
+    this.formFactory = checkNotNull(formFactory);
+  }
+
+  public CompletionStage<Result> create(Request request) {
+    Form<QuestionForm> form = formFactory.form(QuestionForm.class);
+    QuestionForm questionForm = form.bindFromRequest(request).get();
+    return service
+        .getReadOnlyQuestionService()
+        .thenApplyAsync(
+            readOnlyService -> {
+              try {
+                QuestionDefinition definition =
+                    questionForm
+                        .getBuilder()
+                        .setId(readOnlyService.getNextId())
+                        .setVersion(1L)
+                        .build();
+                service.create(definition);
+              } catch (UnsupportedQuestionTypeException e) {
+                // I'm not sure why this would happen here, so we'll just log and redirect.
+                LOG.info(e.toString());
+              }
+              return redirect("/admin/questions");
+            });
+  }
+
+  public CompletionStage<Result> edit(Request request, String path) {
+    return service
+        .getReadOnlyQuestionService()
+        .thenApplyAsync(
+            readOnlyService -> {
+              Optional<QuestionDefinition> definition = Optional.empty();
+              try {
+                definition = Optional.of(readOnlyService.getQuestionDefinition(path));
+              } catch (InvalidPathException e) { // If the path doesn't exist, redirect to create.
+                LOG.info(e.toString());
+              }
+              if (definition.isPresent()) {
+                return ok(editView.render(request, definition));
+              } else {
+                return redirect("/admin/questions/new");
+              }
+            });
+  }
+
+  public Result newOne(Request request) {
+    return ok(editView.render(request, Optional.empty()));
+  }
+
+  public CompletionStage<Result> index(String renderAs) {
```

Comment:
Suggestion (blocking): `renderAs` when it is passed in here to `index` is user input. We should validate that it is one of an allowed set of values and then convert to an enum to pass to the view for rendering on line 95. Alternatively,  we could `switch (renderAs)` and call a different render method on `listView` for each valid value of `renderAs` and throw bad request by default.

Reply:
Thanks! I opted for the switch in the controller.